File: example_py/seer_navigation_example.py
#!/usr/bin/env python3

import argparse
import logging
import math
import signal
import time
from typing import List, Optional

import flexiv_amr


logger = logging.getLogger("seer_navigation_example")
stop_requested = False

NAVIGATION_RUNNING = 2
NAVIGATION_COMPLETED = 4
NAVIGATION_FAILED = 5
NAVIGATION_CANCELED = 6
STATUS_REQUEST_GAP_SECONDS = 0.15
POLL_INTERVAL_SECONDS = 0.35
PAUSE_DURATION_SECONDS = 2.0


def signal_handler(signum, frame) -> None:
    global stop_requested
    stop_requested = True
    logger.info("Stop signal received.")


def finite_float(value: str) -> float:
    parsed = float(value)
    if not math.isfinite(parsed):
        raise argparse.ArgumentTypeError("value must be finite")
    return parsed


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Execute fixed or specified path navigation and monitor status "
            "until completion."
        ),
        epilog=(
            "One target uses API 3051. Multiple targets use API 3066 and must "
            "describe directly connected map lines. This example is for "
            "single-AMR validation, not fleet dispatch."
        ),
    )
    parser.add_argument("amr_ip", help="SEER AMR IPv4 address or hostname")
    parser.add_argument("source_id", help="Source station ID")
    parser.add_argument(
        "target_ids",
        nargs="+",
        help="One or more target station IDs in navigation order",
    )

    parser.add_argument(
        "--pause-after",
        type=finite_float,
        metavar="SECONDS",
        help="Pause once, wait two seconds, then resume",
    )
    parser.add_argument(
        "--timeout",
        type=finite_float,
        default=300.0,
        metavar="SECONDS",
        help="Navigation timeout (default: 300)",
    )

    args = parser.parse_args()
    if args.pause_after is not None and args.pause_after < 0.0:
        parser.error("--pause-after must be non-negative")
    if args.timeout <= 0.0:
        parser.error("--timeout must be greater than zero")
    return args


def sleep_interruptibly(duration_seconds: float) -> None:
    deadline = time.monotonic() + duration_seconds
    while not stop_requested:
        remaining = deadline - time.monotonic()
        if remaining <= 0.0:
            return
        time.sleep(min(remaining, 0.05))


def make_task_id(index: int) -> str:
    return f"seer_navigation_example_{time.time_ns()}_{index}"


def build_navigation_commands(
    source_id: str, target_ids: List[str]
) -> List[flexiv_amr.SeerPathNaviCommand]:
    commands = []
    current_source = source_id

    for index, target_id in enumerate(target_ids):
        command = flexiv_amr.SeerPathNaviCommand()
        command.source_id = current_source
        command.id = target_id
        command.task_id = make_task_id(index)
        commands.append(command)
        current_source = target_id

    return commands


def wait_for_navigation(
    amr: flexiv_amr.SeerAmrClient,
    pause_after_seconds: Optional[float],
    timeout_seconds: float,
) -> str:
    start = time.monotonic()
    deadline = start + timeout_seconds
    pause_exercised = False

    while not stop_requested and time.monotonic() < deadline:
        navigation = amr.get_navigation_status()
        sleep_interruptibly(STATUS_REQUEST_GAP_SECONDS)
        if stop_requested:
            return "interrupted"

        task = amr.get_task_status()
        logger.info(
            "Navigation: status=%s, type=%s, target=%s, finished=%d, unfinished=%d",
            navigation.task_status,
            navigation.task_type,
            navigation.target_id,
            len(navigation.finished_path),
            len(navigation.unfinished_path),
        )
        logger.info(
            "Task: percentage=%.1f, source=%s, target=%s, states=%d, info=%s",
            task.percentage,
            task.source_name,
            task.target_name,
            len(task.task_states),
            task.info,
        )

        if navigation.task_status == NAVIGATION_COMPLETED:
            return "completed"
        if navigation.task_status == NAVIGATION_FAILED:
            return "failed"
        if navigation.task_status == NAVIGATION_CANCELED:
            return "canceled"

        elapsed_seconds = time.monotonic() - start
        if (
            not pause_exercised
            and pause_after_seconds is not None
            and navigation.task_status == NAVIGATION_RUNNING
            and elapsed_seconds >= pause_after_seconds
        ):
            logger.warning("Pausing navigation for two seconds...")
            if not amr.pause_navigation():
                raise RuntimeError("SEER AMR rejected the pause command")
            sleep_interruptibly(PAUSE_DURATION_SECONDS)
            if stop_requested:
                return "interrupted"
            if not amr.resume_navigation():
                raise RuntimeError("SEER AMR rejected the resume command")
            logger.info("Navigation resumed.")
            pause_exercised = True

        sleep_interruptibly(POLL_INTERVAL_SECONDS)

    return "interrupted" if stop_requested else "timed_out"


def main() -> int:
    logging.basicConfig(
        level=logging.INFO,
        format="[%(asctime)s.%(msecs)03d] [%(levelname)s] %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )
    args = parse_arguments()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    client_options = flexiv_amr.SeerAmrOptions()
    client_options.host = args.amr_ip
    client_options.auto_connect = False
    client_options.auto_reconnect = True

    amr = flexiv_amr.SeerAmrClient(client_options)
    navigation_active = False
    connected = False
    exit_code = 1

    try:
        logger.info("Connecting to SEER AMR: %s ...", args.amr_ip)
        amr.connect()
        connected = True
        logger.info("Connected: %s", amr.is_connected())

        emergency = amr.get_emergency_status()
        if emergency.emergency or emergency.driver_emc or emergency.soft_emc:
            raise RuntimeError(
                "AMR emergency stop is active; navigation is not started"
            )

        sleep_interruptibly(STATUS_REQUEST_GAP_SECONDS)
        block = amr.get_block_status()
        if block.blocked:
            logger.warning(
                "AMR is currently blocked, reason=%s, sensor_id=%s",
                block.block_reason,
                block.block_id,
            )

        if stop_requested:
            raise RuntimeError("navigation interrupted before start")

        commands = build_navigation_commands(args.source_id, args.target_ids)
        if len(commands) == 1:
            command = commands[0]
            logger.warning(
                "Starting fixed path navigation: %s -> %s (task_id=%s)",
                command.source_id,
                command.id,
                command.task_id,
            )
            if not amr.execute_fixed_path_navigation(command):
                raise RuntimeError("SEER AMR rejected fixed path navigation")
        else:
            logger.warning(
                "Starting specified path navigation with %d directly connected segments",
                len(commands),
            )
            for command in commands:
                logger.info(
                    "Segment: %s -> %s (task_id=%s)",
                    command.source_id,
                    command.id,
                    command.task_id,
                )
            if not amr.execute_specified_path_navigation(commands):
                raise RuntimeError("SEER AMR rejected specified path navigation")
        navigation_active = True

        sleep_interruptibly(STATUS_REQUEST_GAP_SECONDS)
        result = wait_for_navigation(amr, args.pause_after, args.timeout)
        if result == "completed":
            navigation_active = False
            exit_code = 0
            logger.info("Navigation completed successfully.")
        elif result == "failed":
            navigation_active = False
            logger.error("Navigation failed.")
        elif result == "canceled":
            navigation_active = False
            logger.warning("Navigation was canceled by the controller.")
        elif result == "timed_out":
            logger.error("Navigation timed out.")
        else:
            logger.warning("Navigation interrupted by stop signal.")
    except flexiv_amr.AmrException as exc:
        logger.error("AMR exception: %s", exc)
    except Exception as exc:
        logger.exception("Unexpected exception: %s", exc)
    finally:
        if navigation_active:
            try:
                logger.warning("Canceling active navigation during cleanup...")
                amr.cancel_navigation()
            except Exception as exc:
                logger.exception("Failed to cancel navigation during cleanup: %s", exc)

        if connected:
            try:
                amr.disconnect()
            except Exception as exc:
                logger.exception("Failed to disconnect during cleanup: %s", exc)

    return exit_code


if __name__ == "__main__":
    raise SystemExit(main())

File: example/seer_navigation_example.cpp
/**
 * @example seer_navigation_example.cpp
 * This example executes fixed or specified path navigation and monitors navigation and task status
 * until completion.
 */

#include <chrono>
#include <cmath>
#include <csignal>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

#include <flexiv/amr/exceptions.h>
#include <flexiv/amr/vendor/seer/seer_amr.h>
#include <spdlog/spdlog.h>

namespace {

using flexiv::amr::seer::SeerAmrClient;
using flexiv::amr::seer::SeerPathNaviCommand;

volatile std::sig_atomic_t g_stop_requested = 0;

constexpr int kNavigationRunning = 2;
constexpr int kNavigationCompleted = 4;
constexpr int kNavigationFailed = 5;
constexpr int kNavigationCanceled = 6;
constexpr auto kStatusRequestGap = std::chrono::milliseconds(150);
constexpr auto kPollInterval = std::chrono::milliseconds(350);
constexpr auto kPauseDuration = std::chrono::seconds(2);

struct ExampleOptions {
    std::string amr_ip;
    std::string source_id;
    std::vector<std::string> target_ids;
    std::optional<double> pause_after_seconds;
    double timeout_seconds = 300.0;
};

enum class NavigationResult {
    kCompleted,
    kFailed,
    kCanceled,
    kInterrupted,
    kTimedOut,
};

void SignalHandler(int) {
    g_stop_requested = 1;
}

void PrintUsage(const char* program) {
    std::cerr
        << "Usage:\n"
        << "  " << program << " <amr_ip> <source_id> <target_id> [target_id ...] [options]\n\n"
        << "Options:\n"
        << "  --pause-after <seconds>      Pause once, wait two seconds, then resume.\n"
        << "  --timeout <seconds>          Navigation timeout (default: 300).\n\n"
        << "Examples:\n"
        << "  " << program << " 192.168.192.5 LM1 LM2\n"
        << "  " << program << " 192.168.192.5 LM1 LM2 AP1\n"
        << "  " << program << " 192.168.192.5 LM1 LM2 --pause-after 3\n\n"
        << "One target uses API 3051. Multiple targets use API 3066 and must describe directly "
           "connected map lines. This example is for single-AMR validation, not fleet dispatch.\n";
}

double ParseFiniteDouble(const std::string& text, const char* name) {
    std::size_t parsed = 0;
    const double value = std::stod(text, &parsed);
    if (parsed != text.size() || !std::isfinite(value)) {
        throw std::invalid_argument(std::string(name) + " must be a finite number");
    }
    return value;
}

ExampleOptions ParseOptions(int argc, char* argv[]) {
    if (argc < 4) {
        throw std::invalid_argument("amr_ip, source_id, and at least one target_id are required");
    }

    ExampleOptions options;
    options.amr_ip = argv[1];
    std::vector<std::string> stations;

    for (int i = 2; i < argc; ++i) {
        const std::string argument = argv[i];

        if (argument == "--pause-after") {
            if (i + 1 >= argc) {
                throw std::invalid_argument("--pause-after requires seconds");
            }
            const double seconds = ParseFiniteDouble(argv[++i], "pause-after seconds");
            if (seconds < 0.0) {
                throw std::invalid_argument("pause-after seconds must be non-negative");
            }
            options.pause_after_seconds = seconds;
        } else if (argument == "--timeout") {
            if (i + 1 >= argc) {
                throw std::invalid_argument("--timeout requires seconds");
            }
            options.timeout_seconds = ParseFiniteDouble(argv[++i], "timeout seconds");
            if (options.timeout_seconds <= 0.0) {
                throw std::invalid_argument("timeout seconds must be greater than zero");
            }
        } else if (argument.rfind("--", 0) == 0) {
            throw std::invalid_argument("unknown option: " + argument);
        } else {
            stations.emplace_back(argument);
        }
    }

    if (stations.size() < 2) {
        throw std::invalid_argument("source_id and at least one target_id are required");
    }

    options.source_id = stations.front();
    options.target_ids.assign(stations.begin() + 1, stations.end());
    return options;
}

void SleepInterruptibly(std::chrono::milliseconds duration) {
    constexpr auto kStep = std::chrono::milliseconds(50);
    auto remaining = duration;
    while (!g_stop_requested && remaining > std::chrono::milliseconds::zero()) {
        const auto sleep_duration = remaining < kStep ? remaining : kStep;
        std::this_thread::sleep_for(sleep_duration);
        remaining -= sleep_duration;
    }
}

std::string MakeTaskId(std::size_t index) {
    const auto timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
                               std::chrono::system_clock::now().time_since_epoch())
                               .count();
    return "seer_navigation_example_" + std::to_string(timestamp) + "_" + std::to_string(index);
}

std::vector<SeerPathNaviCommand> BuildNavigationCommands(const ExampleOptions& options) {
    std::vector<SeerPathNaviCommand> commands;
    std::string source_id = options.source_id;

    for (std::size_t index = 0; index < options.target_ids.size(); ++index) {
        SeerPathNaviCommand command;
        command.source_id = source_id;
        command.id = options.target_ids[index];
        command.task_id = MakeTaskId(index);
        commands.emplace_back(std::move(command));
        source_id = options.target_ids[index];
    }

    return commands;
}

NavigationResult WaitForNavigation(SeerAmrClient& amr, const ExampleOptions& options) {
    const auto start = std::chrono::steady_clock::now();
    const auto deadline = start + std::chrono::duration<double>(options.timeout_seconds);
    bool pause_exercised = false;

    while (!g_stop_requested && std::chrono::steady_clock::now() < deadline) {
        const auto navigation = amr.GetNavigationStatus();
        SleepInterruptibly(kStatusRequestGap);
        if (g_stop_requested) {
            return NavigationResult::kInterrupted;
        }

        const auto task = amr.GetTaskStatus();
        spdlog::info("Navigation: status={}, type={}, target={}, finished={}, unfinished={}",
                     navigation.task_status,
                     navigation.task_type,
                     navigation.target_id,
                     navigation.finished_path.size(),
                     navigation.unfinished_path.size());
        spdlog::info("Task: percentage={:.1f}, source={}, target={}, states={}, info={}",
                     task.percentage,
                     task.source_name,
                     task.target_name,
                     task.task_states.size(),
                     task.info);

        if (navigation.task_status == kNavigationCompleted) {
            return NavigationResult::kCompleted;
        }
        if (navigation.task_status == kNavigationFailed) {
            return NavigationResult::kFailed;
        }
        if (navigation.task_status == kNavigationCanceled) {
            return NavigationResult::kCanceled;
        }

        const double elapsed_seconds =
            std::chrono::duration<double>(std::chrono::steady_clock::now() - start).count();
        if (!pause_exercised && options.pause_after_seconds.has_value() &&
            navigation.task_status == kNavigationRunning &&
            elapsed_seconds >= options.pause_after_seconds.value()) {
            spdlog::warn("Pausing navigation for two seconds...");
            if (!amr.PauseNavigation()) {
                throw std::runtime_error("SEER AMR rejected the pause command");
            }

            SleepInterruptibly(
                std::chrono::duration_cast<std::chrono::milliseconds>(kPauseDuration));
            if (g_stop_requested) {
                return NavigationResult::kInterrupted;
            }

            if (!amr.ResumeNavigation()) {
                throw std::runtime_error("SEER AMR rejected the resume command");
            }
            spdlog::info("Navigation resumed.");
            pause_exercised = true;
        }

        SleepInterruptibly(kPollInterval);
    }

    return g_stop_requested ? NavigationResult::kInterrupted : NavigationResult::kTimedOut;
}

}  // namespace

int main(int argc, char* argv[]) {
    if (argc >= 2 && (std::string(argv[1]) == "--help" || std::string(argv[1]) == "-h")) {
        PrintUsage(argv[0]);
        return 0;
    }

    ExampleOptions example_options;
    try {
        example_options = ParseOptions(argc, argv);
    } catch (const std::exception& e) {
        std::cerr << "Invalid arguments: " << e.what() << "\n\n";
        PrintUsage(argv[0]);
        return 1;
    }

    std::signal(SIGINT, SignalHandler);
    std::signal(SIGTERM, SignalHandler);
    spdlog::set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%^%l%$] %v");

    flexiv::amr::seer::SeerAmrOptions client_options;
    client_options.host = example_options.amr_ip;
    client_options.auto_connect = false;
    client_options.auto_reconnect = true;

    SeerAmrClient amr(client_options);
    bool connected = false;
    bool navigation_active = false;
    int exit_code = 1;

    try {
        spdlog::info("Connecting to SEER AMR: {} ...", example_options.amr_ip);
        amr.Connect();
        connected = true;
        spdlog::info("Connected: {}", amr.IsConnected());

        const auto emergency = amr.GetEmergencyStatus();
        if (emergency.emergency || emergency.driver_emc || emergency.soft_emc) {
            throw std::runtime_error("AMR emergency stop is active; navigation is not started");
        }

        SleepInterruptibly(kStatusRequestGap);
        const auto block = amr.GetBlockStatus();
        if (block.blocked) {
            spdlog::warn("AMR is currently blocked, reason={}, sensor_id={}",
                         block.block_reason,
                         block.block_id);
        }

        if (g_stop_requested) {
            throw std::runtime_error("navigation interrupted before start");
        }

        const auto commands = BuildNavigationCommands(example_options);
        if (commands.size() == 1) {
            spdlog::warn("Starting fixed path navigation: {} -> {} (task_id={})",
                         commands.front().source_id.value(),
                         commands.front().id.value(),
                         commands.front().task_id.value());
            if (!amr.ExecuteFixedPathNavigation(commands.front())) {
                throw std::runtime_error("SEER AMR rejected fixed path navigation");
            }
        } else {
            spdlog::warn("Starting specified path navigation with {} directly connected segments",
                         commands.size());
            for (const auto& command : commands) {
                spdlog::info("Segment: {} -> {} (task_id={})",
                             command.source_id.value(),
                             command.id.value(),
                             command.task_id.value());
            }
            if (!amr.ExecuteSpecifiedPathNavigation(commands)) {
                throw std::runtime_error("SEER AMR rejected specified path navigation");
            }
        }
        navigation_active = true;

        SleepInterruptibly(kStatusRequestGap);
        const NavigationResult result = WaitForNavigation(amr, example_options);
        if (result == NavigationResult::kCompleted) {
            navigation_active = false;
            exit_code = 0;
            spdlog::info("Navigation completed successfully.");
        } else if (result == NavigationResult::kFailed) {
            navigation_active = false;
            spdlog::error("Navigation failed.");
        } else if (result == NavigationResult::kCanceled) {
            navigation_active = false;
            spdlog::warn("Navigation was canceled by the controller.");
        } else if (result == NavigationResult::kTimedOut) {
            spdlog::error("Navigation timed out.");
        } else {
            spdlog::warn("Navigation interrupted by stop signal.");
        }
    } catch (const flexiv::amr::AmrException& e) {
        spdlog::error("AMR exception: {}, code: {}", e.what(), static_cast<int>(e.code()));
    } catch (const std::exception& e) {
        spdlog::error("std exception: {}", e.what());
    }

    if (navigation_active) {
        try {
            spdlog::warn("Canceling active navigation during cleanup...");
            amr.CancelNavigation();
        } catch (const std::exception& e) {
            spdlog::error("Failed to cancel navigation during cleanup: {}", e.what());
        }
    }

    if (connected) {
        try {
            amr.Disconnect();
        } catch (const std::exception& e) {
            spdlog::error("Failed to disconnect during cleanup: {}", e.what());
        }
    }

    return exit_code;
}
